Rework memory layout

andrea-iob opened the pull request:
Data needed for the computation are now stored using structures of array. SoA should allow the GPU to make coalesced memory acceses.

Also, the containers that store device data are passed down to the low level kernel functions. In this way it is not necessary to gather all the data needed by low level functions before calling the functions. Instead, each function will access the data only when it actually needs. This reduces the amount of memory stalls because, since data access and computation are interleaved, the GPU is able to access the data needed later on while doing some computations.

@kgkara @marcocisternino Can someone profile the code to check if the get some real speedups?

andrea-iob committed:
containers: explicitly declare storage classes also when CUDA is not enabled

andrea-iob committed:
minimmerflow: rework memory layout

andrea-iob committed:
containers: fix include guard

andrea-iob committed:
containers: add objects for defining cursors to device data

andrea-iob committed:
minimmerflow: try to interleave memory accesses and computation in CUDA kernels

This can be achieved passing the containers that store device data
to the low level kernel functions. In this way it is not necessary
to gather all the data needed by low level functions before calling
the functions. Instead, each function will access the data only when
it actually needs. This reduces the amount of memory stalls because,
since data access and computation are interleaved, the GPU is able to
access the data needed later on while doing some computations.

marcocisternino commented:
I’m going to profile it and add the result to the presentation

Marco Cisternino, PhD
***@***.******@***.***>
______________________
Optimad Engineering Srl
Via Bligny 5, Torino, Italia.
+3901119719782
www.optimad.it<http://www.optimad.it/>

From: Andrea Iob ***@***.***>
Sent: lunedì 22 novembre 2021 08:40
To: optimad/minimmerflow ***@***.***>
Cc: Marco Cisternino ***@***.***>; Mention ***@***.***>
Subject: [optimad/minimmerflow] Rework memory layout (PR #6)


Data needed for the computation are now stored using structures of array. SoA should allow the GPU to make coalesced memory acceses.

Also, the containers that store device data are passed down to the low level kernel functions. In this way it is not necessary to gather all the data needed by low level functions before calling the functions. Instead, each function will access the data only when it actually needs. This reduces the amount of memory stalls because, since data access and computation are interleaved, the GPU is able to access the data needed later on while doing some computations.

@kgkara<https://github.com/kgkara> @marcocisternino<https://github.com/marcocisternino> Can someone profile the code to check if the get some real speedups?

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/optimad/minimmerflow/pull/6

Commit Summary

  *   34230f7<https://github.com/optimad/minimmerflow/pull/6/commits/34230f7ec087216f2e24aa931b6221c19d529cee> containers: add containers to handle collection of storages
  *   59c0be9<https://github.com/optimad/minimmerflow/pull/6/commits/59c0be9ba517aad1ad542a17beb8de85ccb670b8> containers: use explicit instantiation also when CUDA support is disabled
  *   245615f<https://github.com/optimad/minimmerflow/pull/6/commits/245615f7f12dd2d911bd31f0054aaed27b042c7a> containers: fix include guard
  *   b678f45<https://github.com/optimad/minimmerflow/pull/6/commits/b678f459bb2b8304f386f54559a0e38d10adabbb> containers: add objects for defining cursors to device data
  *   7a81e72<https://github.com/optimad/minimmerflow/pull/6/commits/7a81e728e411090a8603811433edc70d267ae118> minimmerflow: rework memory layout
  *   84830d0<https://github.com/optimad/minimmerflow/pull/6/commits/84830d0d40bb234f2bd6b100c3ae5e3c1e76592e> minimmerflow: try to interleave memory accesses and computation in CUDA kernels

File Changes

(24 files<https://github.com/optimad/minimmerflow/pull/6/files>)

  *   M src/computation_info.cpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-0524bb965175e13e17b4a1a2a80429617dc1d7a79a3ac0f3f04ba0b5907c8fb9> (18)
  *   M src/computation_info.cu<https://github.com/optimad/minimmerflow/pull/6/files#diff-e6232cccdad4c99ae454da14f11930d1aabbf15d8f56d7c886877a2ea2e1c925> (19)
  *   M src/computation_info.hpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-0f35e372346c35b5351f2902b6169c306686591086a525d25606dfb511f25ec7> (16)
  *   A src/containers.cpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-323b0044b4667baf7605d69780d975a5860d7e99c24e2f25a2598fd4d36c13e7> (68)
  *   M src/containers.cu<https://github.com/optimad/minimmerflow/pull/6/files#diff-6e3d2167a487ace4037dd02baf4c2ce1b30469537c4ec7318d0b73c605969c07> (45)
  *   M src/containers.hcu<https://github.com/optimad/minimmerflow/pull/6/files#diff-f3ae0446c740abf2837cd0dd3e74a7b004b8c929f3e5e723e47f144b1462b97b> (87)
  *   M src/containers.hpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-73fdea5cb57fdbe82d0906d069b2396eefe048ff1a97e884709cecade035216a> (120)
  *   M src/containers.tcu<https://github.com/optimad/minimmerflow/pull/6/files#diff-836bbde7deb9e16cbd8943a370438ded9c8fa346c2a137d4094ede460887649e> (232)
  *   A src/containers.tpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-f6397d877fdfcdea2a86c73b1761559d25ec9a0402f683e8bf2b38f3875f8efc> (91)
  *   M src/euler.cpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-46d406d34b286f0989aef8d9801adcd856e8ea42bf1e06d44d82f30e4ff51f1d> (43)
  *   M src/euler.cu<https://github.com/optimad/minimmerflow/pull/6/files#diff-69fcf87a666643a592326006d5d13df58cb1d45bfe25f53a92d5fbd0e90b2069> (190)
  *   M src/euler.hcu<https://github.com/optimad/minimmerflow/pull/6/files#diff-e088b556005c6651e8e46258e1be9dd07f8ed3af3585d39aabc5df5817b3b516> (73)
  *   M src/euler.hpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-5e52a432067572f2610a5261184a07cd4cbc4ecb024bc69ce06163bb85354840> (10)
  *   M src/main.cpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-34d21af3c614ea3cee120df276c9c4ae95053830d7f1d3deaf009a4625409ad2> (143)
  *   M src/problem.cu<https://github.com/optimad/minimmerflow/pull/6/files#diff-90c05af11795b7d0fcf89d805b972873a097739c92c5ac31a2901269447dbf5f> (5)
  *   M src/problem.hcu<https://github.com/optimad/minimmerflow/pull/6/files#diff-8a9b46ce456c9196ee3eb28babf3e82b34409271b7f0567ba09029437026d119> (6)
  *   M src/reconstruction.cpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-6ca8fe251d7a7bec5bacbf261dfd957c42301d06a8bb4eb987f717d99da86af4> (2)
  *   M src/reconstruction.cu<https://github.com/optimad/minimmerflow/pull/6/files#diff-0e3b7a36da0732ecf6130ef02299f8862fc322e50da0fa672c94ce74f9d8582f> (8)
  *   M src/reconstruction.hcu<https://github.com/optimad/minimmerflow/pull/6/files#diff-e6677ed1fae61851ee560f32047be5db1deb163086f1634a1957374850e8e252> (7)
  *   M src/reconstruction.hpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-ee9b75b277227ec671e703ba2d778e60d41292dec8750eb61812e7d93bff1bd1> (2)
  *   M src/solver_writer.cpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-2614f35b1768a625fc32096d66a75996f5ff772409426bb9196bd0eeeab54b6b> (10)
  *   M src/solver_writer.hpp<https://github.com/optimad/minimmerflow/pull/6/files#diff-d848708a317a28caf538ddfa18906f8d958bf67c76293f0b7fadf14ddaec349c> (10)
  *   M src/utils.cu<https://github.com/optimad/minimmerflow/pull/6/files#diff-2e7399b37465b7ac17d131be2a5c1fd14e4c74633b4809b20c12407cd17dd85c> (41)
  *   M src/utils.hcu<https://github.com/optimad/minimmerflow/pull/6/files#diff-8922590c567d8bc0f1a906447b270d01364ca8324d14c68db77cf4af572efadd> (9)

Patch Links:

  *   https://github.com/optimad/minimmerflow/pull/6.patch
  *   https://github.com/optimad/minimmerflow/pull/6.diff

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/optimad/minimmerflow/pull/6>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABSMIGAOXCL2PV5HNPG2ZSLUNHXVHANCNFSM5IQJICHQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

andrea-iob commented:
The main issue with the changes was that the [] operator was a virtual function. I fixed this and added some optimizations to the evaluation of primitive/conservative values (e.g., evaluate terms like 1 / rho, u * u once), but, even with this optimizations, the usage of SoA is slower than the original AoS code (1.4ms vs 1.2ms). 

During this week I will trying one more thing: refactor the interface to allow coalesced access. Let's see if, removing this last bottleneck, we see some differences.

(Right now this branch is a little bit messy, I will tidy it up during this week.)

andrea-iob commented:
Note: I've also removed the extra kinetic energy evaluation, but the speedup was negligible.

andrea-iob committed:
mesh: store vector data using a ScalarPiercedStorageCollection

andrea-iob committed:
utils: force functions to be inlined

andrea-iob committed:
containers: add containers to handle collection of storages

andrea-iob commented:
Uncoalesced accesses are now gone. I've also improved a little bit more the calculations of the fluxes. Here the latest results:

![ncu](https://user-images.githubusercontent.com/7443174/143314866-646d73a8-3011-47b6-8323-b208123eccfa.png)

The speedup seems to come form the optimizations to the calculation of the fluxes rather than from the removal of the uncoalesced accesses.

andrea-iob committed:
problem: cosmetix fixes

MircoValentini commented:
I agree that the speedup comes from the optimization of the flux computation (the same behaviour can be inferred from the unrolled branches). With just 5 variables the problems arising from an uncoaleshed access are  hidden by the by the very effective work of the L1 cache. But as the number of variable increase, the difference between the two approaches (coaleshed and uncoaleshed) will increase also (it can be "somewhat" inferred from the unrolled.strided example)

andrea-iob committed:
minimmerflow: use shared memory in residual update

andrea-iob committed:
reconstructions: force functions to be inlined

andrea-iob committed:
euler: micro optimize evaluation of interface eigenvalue

andrea-iob committed:
euler: micro optimize evaluation of interface fluxes

andrea-iob committed:
utils: simplify evaluation of convervative/primitive variables

andrea-iob committed:
containers: add data structures for handling data stored in shared memory

andrea-iob committed:
containers: use double qoutes when including local headers

andrea-iob committed:
euler: rename dev_evalSplitting to dev_solveRiemann

andrea-iob commented:
Using shared memory and inlining evaluation of reconstructions we can achieve a 2.9x speedup with respect to the original version.

When doing the comparison with the original version, we need to consider that the new kernel does both flux evaluation and interface reconstructions (two reconstruction for each interface). In the original version, the time spent in each reconstruction was about 0.45ms (0.9ms in total), this should be summed to the time spent in fluxes evaluation. All considered the new kernel takes 0.74 ms, whereas the old one took (0.45 + 0.45 + 1.24) = 2.14 ms.

![inline](https://user-images.githubusercontent.com/7443174/143883534-000469b4-5299-44c8-919a-1d8a6d80c61e.png)
  All considered there is a 2.25x speedup.

andrea-iob committed:
containers: prevent changing the stride of a DeviceProxyArray

andrea-iob committed:
euler: optimize evaluation of fluxes

andrea-iob committed:
containers: define collection data also on the host

andrea-iob commented:
Instead of merging the changes into cuda.v2 branch, the rework of the memory layout will be part of a new CUDA revision (branch cuda.v3).
